### PR TITLE
Colocate Join (#245)

### PR DESCRIPTION
**Pull request by kangkaisen**
[For 245](https://github.com/apache/incubator-doris/issues/245)

Please review, Thanks!

**About Test**
I will add test later.

### CODE REVIEW

File: fe/src/com/baidu/palo/qe/SessionVariable.java
Comment (imay) [inline fe/src/com/baidu/palo/qe/SessionVariable.java]: { should in the same line
public boolean isDisableColocateJoin() {

File: gensrc/thrift/PaloInternalService.thrift
Comment (imay) [inline gensrc/thrift/PaloInternalService.thrift]: This flag is useless to BE, no need to add this option

File: fe/src/com/baidu/palo/planner/HashJoinNode.java
Comment (imay) [inline fe/src/com/baidu/palo/planner/HashJoinNode.java]: code style

File: fe/src/com/baidu/palo/planner/DistributedPlanner.java
Comment (imay) [inline fe/src/com/baidu/palo/planner/DistributedPlanner.java]: code style: 
should be one empty line

File: fe/src/com/baidu/palo/alter/DecommissionBackendJob.java
Comment (imay) [inline fe/src/com/baidu/palo/alter/DecommissionBackendJob.java]: LOG.info will produce too many log? 

change this to debug? or drop this log?

File: fe/src/com/baidu/palo/catalog/Catalog.java
Comment (imay) [inline fe/src/com/baidu/palo/catalog/Catalog.java]: code style: 
4 space indentation

File: fe/src/com/baidu/palo/clone/BackendInfo.java
Comment (imay) [inline fe/src/com/baidu/palo/clone/BackendInfo.java]: code style

**Review by imay — CHANGES_REQUESTED** [review]
some code style case

The code is a bit more, take a closer look at the logic later.

**Comment by kangkaisen** [comment]
Update this commit. 

Because the  change for latest code is big, I will do a full test on this weekend.

**Review by imay — DISMISSED** [review]
You should update your commit log, change it to like "support colocate join" or something, And in the commit log body, you can add more detail about this feature. 

And there are code conflicts to be resolved

**Comment by kangkaisen** [comment]
@imay  I have Updated the commit message and resolved the conflicts. please review. Thanks.

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Currently, alter table colocate_with property need global ALTER Privilege.  so normal user  couldn't alter and we shouldn't grant  normal user  global ALTER Privilege.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: Yes. OK.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: OK

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: OK

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: Yes. I see.

I could merge `logColocateAddTable` and `logColocateBackendsPerBucketSeq` to one ,but `createTableWithLock` still have on log. how do we handle this better ?

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: 1 schemaChangeHandler.process has kept  the db write lock, so we needn't add lock here.

2 OK, I will add a check

File: fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java]: OK.

File: fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/DistributedPlanner.java]: OK

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Why? If one user have ALTER privilege for two tables, it's reasonable that he could make these two tables colocated.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (imay) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: We must keep metadata keep consistent. If one operation failed, we should guarantee that thing can still work.

Following is a suggest.

If we `createTableWithLock ` first, and log another operation `logChangeTableToColocate`.  And     if the second log failed, we also have complete meta, which have a table without colocate property. 

Can this method work and be implemented easily.

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Yes.  it's reasonable user could later table colocate_with property. but require a global ALTER Privilege is not reasonable. normal users shouldn't have global ALTER Privilege, if they have, they will could read all dbs.  I tested it. So we should check db level Privilege not global Privilege.

File: fe/src/main/java/org/apache/doris/catalog/Catalog.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/catalog/Catalog.java]: I see. OK

File: fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java
Comment (imay) [reply fe/src/main/java/org/apache/doris/analysis/ModifyTablePropertiesClause.java]: Yes, you're right, I mistakenly thought that the DB privilege was checked before.

**Comment by imay** [comment]
@kangkaisen 
Could you add some `help doc` to this PR, which help user to use this feature.

**Comment by kangkaisen** [comment]
OK, I will add design doc  and help doc later.

**Comment by kangkaisen** [comment]
Add help doc for colocate join.

**Comment by kangkaisen** [comment]
Hi. @imay Is there any question about this PR?  If not, I will resolve the conflict and rebase the commit.

**Comment by imay** [comment]
> Hi. @imay Is there any question about this PR? If not, I will resolve the conflict and rebase the commit.

Some things have hindered me, so I haven't reviewed yet. I will reply to you earlier next week?

**Comment by kangkaisen** [comment]
OK，Thanks you.

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/clone/ColocateTableBalancer.java]: OK

File: fe/src/main/java/org/apache/doris/planner/OlapScanNode.java
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/planner/OlapScanNode.java]: OK

**Comment by kangkaisen** [comment]
Add more comment for `ColocateTableBalancer`

**Comment by kangkaisen** [comment]
Rebase this commit.

**Review by morningman — APPROVED** [review]
LGTM, but I need a further review after merged

**Review by imay — APPROVED** [review]

**Pull request merged**
